src/java.base/share/classes/jdk/internal/reflect/ReflectionFactory.java

```diff
@@ -623,74 +588,152 @@ public final boolean hasStaticInitializerForSerialization(Class<?> cl) {
+        private static final int METHOD_MH_ACCESSOR      = 0x1;
+        private static final int FIELD_MH_ACCESSOR       = 0x2;
+        private static final int ALL_MH_ACCESSORS        = METHOD_MH_ACCESSOR|FIELD_MH_ACCESSOR;
+
+        private final int     useDirectMethodHandle;
+        private final boolean useNativeAccessorOnly;  // for testing only
+
+        // true if deserialization constructor checking is disabled
+        private final boolean disableSerialConstructorChecks;
+
+        private static final Config DEFAULT = new Config(
+                false, // noInflation
+                15, // inflationThreshold
+                ALL_MH_ACCESSORS, // useDirectMethodHandle
+                false, // useNativeAccessorOnly
+                false // disableSerialConstructorChecks
+        );
+
+        /** We have to defer full initialization of this class until after
+         the static initializer is run since java.lang.reflect.Method's
+         static initializer (more properly, that for
+         java.lang.reflect.AccessibleObject) causes this class's to be
+         run, before the system properties are set up. */
+        private static @Stable Config instance;
+
+        private Config(boolean noInflation, int inflationThreshold, int useDirectMethodHandle,
+                       boolean useNativeAccessorOnly, boolean disableSerialConstructorChecks) {
+            this.noInflation = noInflation;
+            this.inflationThreshold = inflationThreshold;
+            this.useDirectMethodHandle = useDirectMethodHandle;
+            this.useNativeAccessorOnly = useNativeAccessorOnly;
+            this.disableSerialConstructorChecks = disableSerialConstructorChecks;
         }
 
-        val = props.getProperty("sun.reflect.inflationThreshold");
-        if (val != null) {
-            try {
-                inflationThreshold = Integer.parseInt(val);
-            } catch (NumberFormatException e) {
-                throw new RuntimeException("Unable to parse property sun.reflect.inflationThreshold", e);
+        private static Config instance() {
+            // Defer initialization until module system is initialized so as
+            // to avoid inflation and spinning bytecode in unnamed modules
+            // during early startup.
+            if (!VM.isModuleSystemInited()) {
+                return DEFAULT;
             }
-        }
-        val = props.getProperty("jdk.reflect.useDirectMethodHandle");
-        if (val != null) {
-            if (val.equals("false")) {
-                useDirectMethodHandle = 0;
-            } else if (val.equals("methods")) {
-                useDirectMethodHandle = METHOD_MH_ACCESSOR;
-            } else if (val.equals("fields")) {
-                useDirectMethodHandle = FIELD_MH_ACCESSOR;
+
+            Config c = instance;
+            if (c == null) {
+                instance = c = load();
             }
+            return c;
```

Comment:
If you do that the "old" way, you loose the ability for JIT to constant-fold the `instance` and by transitivity the Config instance fields, since the check for `VM.isModuleSystemInited()` can't be elided. As suggested, the fast-path check should be done 1st, like:

```java
    private static @Stable Config instance;

    private static Config instance() {
        Config c = instance;
        if (c != null) {
            return c;
        }
        
        // Defer initialization until module system is initialized so as
        // to avoid inflation and spinning bytecode in unnamed modules
        // during early startup.
        if (!VM.isModuleSystemInited()) {
            return DEFAULT;
        }

        instance = c = load();
        return c;
    }
```

Reply:
...having suggested that rearrangement, perhaps the right way to do it is to enable some VM.isXXX queries themselves to be constant-foldable so that other callers too may benefit. Like this:
https://github.com/plevart/jdk/commit/e918ccc52bbc288f6721af5fa66d8f7a8cc880cf
WDYT?

Reply:
I believe your patch to fold these methods is a good choice: for example, `FileSystems.getDefault()` will be constant-foldable as a result.
For shutdown, the benefit may look negligible, but a consistency in style is beneficial.
To make this more efficient, I recommend looking at the callers to `VM.initLevel()` and replace with such boolean checks if possible: for example, `ClassLoader.getSystemClassLoader` may be constant-foldable if its default branch of switch on init level become a dedicated fast path.

Since this change affects multiple components and beyond the reflection factory itself, I don't think I will include it here; I will just use the right arrangement.

Reply:
Right, I wasn't suggesting to include that in the patch. Local rearrangement is OK anyway. Latest code looks good to me.